### PR TITLE
Add support to Polygon (Matic) chain

### PR DESCRIPTION
**Pull request by py-zoid**
## Description

Add support for the Matic chain.

Matic is a modified Ethereum fork. Hence the Ethereum signing logic is reused. The integration is very similar to the SmartChain integration.

## Testing instructions

Ethereum tests are reused. Added Matic to Ethereum tests similar to how SmartChain did.

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: coins.json [hunk 1]
```diff
@@ -1543,5 +1543,29 @@
       "clientPublic": "https://bsc-dataseed1.binance.org",
       "clientDocs": "https://eth.wiki/json-rpc/API"
     }
+  },
+  {
+    "id": "matic",
```
Comment (hewigovens) [inline coins.json:1548]: let's use new name `polygon` and new symbol

File: include/TrustWalletCore/TWCoinType.h [hunk 2]
```diff
@@ -85,6 +85,7 @@ enum TWCoinType {
     TWCoinTypeSmartChainLegacy = 10000714,
     TWCoinTypeSmartChain = 20000714,
     TWCoinTypeOasis = 474,
+    TWCoinTypeMatic = 966,
```
Comment (hewigovens) [inline include/TrustWalletCore/TWCoinType.h:88]:
```suggestion
    TWCoinTypePolygon = 966,
```

File: coins.json (hunk 1 above)
Comment (py-zoid) [reply coins.json:1548]: Our Coin name and symbol are still Matic and 'MATIC' respectively. It's only the company name that changed.

**Review by catenocrypt — CHANGES_REQUESTED** [review]
Please add/extend ios&android tests, see:
https://github.com/trustwallet/wallet-core/blob/master/swift/Tests/Blockchains/BinanceSmartChainTests.swift
https://github.com/trustwallet/wallet-core/blob/master/swift/Tests/CoinAddressDerivationTests.swift#L73
https://github.com/trustwallet/wallet-core/blob/master/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/CoinAddressDerivationTests.kt#L39
https://github.com/trustwallet/wallet-core/blob/master/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/smartchain/TestBinanceSmartChainAddress.kt

Otherwise LGTM

File: coins.json (hunk 1 above)
Comment (vikmeup) [reply coins.json:1548]: `id` is the project name, not related to the coin itself. Polygon would be a proper name for it

Comment (py-zoid) [reply coins.json:1548]: Thanks for clarifying. I'll change the id to "polygon". Should I have to regenerate the files again after changing the id? Also should I keep the name as Matic or change it to Polygon?

Comment (hewigovens) [reply coins.json:1548]: Yes; and Is Matic a native token on side chain like BNB?

Comment (py-zoid) [reply coins.json:1548]: Yes, Matic is the native token.

**Comment by catenocrypt** [comment]
For completeness, new matic is still missing in these 2 places, please add.

https://github.com/trustwallet/wallet-core/blob/master/swift/Tests/CoinAddressDerivationTests.swift#L73
https://github.com/trustwallet/wallet-core/blob/master/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/CoinAddressDerivationTests.kt#L39

**Comment by py-zoid** [comment]
@catenocrypt The two changes you asked for have already been added in commit fae08c60172d8dae17eb63e477b48cdec451214f ( Modified Address, Signer and Coin tests).

**Comment by catenocrypt** [comment]
> @catenocrypt The two changes you asked for have already been added in commit [fae08c6](https://github.com/trustwallet/wallet-core/commit/fae08c60172d8dae17eb63e477b48cdec451214f) ( Modified Address, Signer and Coin tests).

Oh, OK great, sorry!

**Review by catenocrypt — APPROVED** [review]

File: coins.json (hunk 1 above)
Comment (hewigovens) [reply coins.json:1548]: from `https://polygon.technology/get-started/`, are you adding `Matic PoS` or `Matic Plasma`?

Comment (py-zoid) [reply coins.json:1548]: Currently Matic Network is one single POS chain. For exit mechanism we use Plasma and POS bridges.

File: coins.json [hunk 3]
```diff
@@ -1543,5 +1543,29 @@
...
+    "name": "Matic",
```
Comment (hewigovens) [inline coins.json:1549]: From https://coinmarketcap.com/currencies/polygon/
```suggestion
    "name": "Polygon",
```

**Review by hewigovens — CHANGES_REQUESTED** [review]
let's at least change name to match CMC?

**Comment by py-zoid** [comment]
Changed the id to "polygon" and name to "Polygon". Let me know if there are anymore changes to be made

**Review by hewigovens — APPROVED** [review]

**Comment by sggts04** [comment]
Hi, I am under the impression that this merged PR added support for Polygon network, however I still don't see an option for Matic/Polygon chain in Trust Wallet App. Am I missing something?

**Comment by sggts04** [comment]
Hi, I am under the impression that this merged PR added support for Polygon network, however I still don't see an option for Matic/Polygon chain in Trust Wallet App. Am I missing something? @hewigovens

**Pull request merged**